### PR TITLE
Bugfix: avert endless loop on unexpected EOF at ipc messages

### PR DESCRIPTION
**Pull request by xzfc**
Fix freeze on invalid ipc commands like

    echo -n $'i3-ipc\0\0\0\xa\0\0\0\0focus left' | socat - `i3 --get-socketpath`

Also, treat incomplete headers as IPC violation. Example of incomplete header:

    echo -n i3-ip | socat - `i3 --get-socketpath`

### CODE REVIEW

File: libi3/ipc_recv_message.c [hunk 1]
```diff
@@ -41,7 +41,12 @@ int ipc_recv_message(int sockfd, uint32_t *message_type,
         if (n == -1)
             return -1;
         if (n == 0) {
-            return -2;
+            if (read_bytes == 0) {
+                return -2;
+            } else {
+                ELOG("IPC: unexpected EOF\n");
```
Comment (stapelberg) [inline libi3/ipc_recv_message.c]: Here and below, let’s make the error messages a bit more helpful. How about:

```c
ELOG("IPC: unexpected EOF after %d bytes, want %d bytes\n", read_bytes, to_read);
```

File: libi3/ipc_recv_message.c [hunk 2]
```diff
@@ -61,13 +66,17 @@ int ipc_recv_message(int sockfd, uint32_t *message_type,
     *reply = smalloc(*reply_length);
 
     read_bytes = 0;
-    int n;
     while (read_bytes < *reply_length) {
-        if ((n = read(sockfd, *reply + read_bytes, *reply_length - read_bytes)) == -1) {
+        int n = read(sockfd, *reply + read_bytes, *reply_length - read_bytes);
```
Comment (stapelberg) [inline libi3/ipc_recv_message.c]: Make this `const int n` please

**Comment by xzfc** [comment]
Updated.

File: libi3/ipc_recv_message.c (hunk 1 above)
Comment (xzfc) [reply libi3/ipc_recv_message.c]: I've rephrased slightly to distinguish EOF in payload/header:

```
IPC: unexpected EOF while reading payload, got 10 bytes, want 167772160 bytes
IPC: unexpected EOF while reading header, got 5 bytes, want 14 bytes
IPC: invalid magic in header, got "i3-ppc", want "i3-ipc"
```

**Comment by stapelberg** [comment]
Neat, thanks!

**Pull request merged**